nntrainer/layers/multi_head_attention_layer.cpp

```diff
@@ -55,28 +54,257 @@ enum AttentionParams {
+  const float &bias_decay = std::get<props::BiasDecay>(*layer_impl_props).get();
+
+  NNTR_THROW_IF(std::get<props::NumHeads>(multi_head_attention_props).empty(),
+                std::invalid_argument)
+    << "num_heads property is not provided for layer " << context.getName();
+  const unsigned int num_heads =
+    std::get<props::NumHeads>(multi_head_attention_props).get();
+
+  if (std::get<props::ProjectedKeyDim>(multi_head_attention_props).empty()) {
+    NNTR_THROW_IF(query_width % num_heads, std::invalid_argument)
+      << "query_width: " << query_width
+      << " is not divisible by num_heads: " << num_heads << " for layer "
+      << context.getName();
+
+    ml_logw("[multi head attention] ProjectedKeyDim property is not given. "
+            "Default value(query_width / num_heads) is set");
+
+    std::get<props::ProjectedKeyDim>(multi_head_attention_props)
+      .set(query_width / num_heads);
+  }
+  const unsigned int projected_key_dim_prop =
+    std::get<props::ProjectedKeyDim>(multi_head_attention_props).get();
+
+  if (std::get<props::ProjectedValueDim>(multi_head_attention_props).empty()) {
+    std::get<props::ProjectedValueDim>(multi_head_attention_props)
+      .set(projected_key_dim_prop);
+  }
+  const unsigned int projected_value_dim_prop =
+    std::get<props::ProjectedValueDim>(multi_head_attention_props).get();
+
+  if (std::get<props::OutputShape>(multi_head_attention_props).empty()) {
+    std::get<props::OutputShape>(multi_head_attention_props).set(query_width);
+  }
+  const unsigned int output_shape =
+    std::get<props::OutputShape>(multi_head_attention_props).get();
+
+  const float dropout_rate =
+    std::get<props::DropOutRate>(multi_head_attention_props).get();
+
+  if (std::get<props::AverageAttentionWeight>(multi_head_attention_props)
+        .empty()) {
+    std::get<props::AverageAttentionWeight>(multi_head_attention_props)
+      .set(true);
+  }
+  const bool average_attention_weight =
+    std::get<props::AverageAttentionWeight>(multi_head_attention_props).get();
+
+  const props::ReturnAttentionWeightInfo::Enum return_attention_weight =
+    std::get<props::ReturnAttentionWeight>(multi_head_attention_props).get();
+
+  const unsigned int projected_query_dim_prop = projected_key_dim_prop;
+
+  sm.setActiFunc(ActivationType::ACT_SOFTMAX);
+
+  NNTR_THROW_IF(query_dim.channel() != 1, std::invalid_argument)
+    << "Dimension of input query channel: " << query_dim.channel()
+    << " is not 1 for layer " << context.getName();
+  NNTR_THROW_IF(key_dim.channel() != 1, std::invalid_argument)
+    << "Dimension of input key channel: " << key_dim.channel()
+    << " is not 1 for layer " << context.getName();
+  NNTR_THROW_IF(value_dim.channel() != 1, std::invalid_argument)
+    << "Dimension of input value channel: " << value_dim.channel()
+    << " is not 1 for layer " << context.getName();
+  NNTR_THROW_IF(provide_attention_mask && mask_dim.channel() != num_heads,
```

Comment:
How about extect ```provide_attention_mask``` to other throw ?
line 165, 175, 181 are using it.

Reply:
In multi head attention the input mask might not given so before checking the mask dimension checking the mask is provided or not is needed.
So I think we can't extract provide_attention_mask to other throw.

Reply:
Oh, I see. thanks for checking that.